Fail fast when a new database transaction cannot be started.

trajber opened the pull request:
We need a new approach to handle database failures. Since db.Begin() blocks
when the database server is not responding, the current implementation, in the
worst case, sleeps a given duration for each request. This cannot be an option;
it can exhaust server resources quickly under high loads.  This commit
implements a simple solution that starts a goroutine to check database
periodically. When the database server still freezed all the following requests
will fail and will not wait sleeping.
@joaopetreli @rafaeljusto

Commit:
Fail fast when a new database transaction cannot be started.

We need a new approach to handle database failures. Since db.Begin() blocks
when the database server is not responding, the current implementation, in the
worst case, sleeps a given duration for each request. This cannot be an option;
it can exhaust server resources quickly under high loads.  This commit
implements a simple solution that starts a goroutine to check database
periodically. When the database server still freezed all the following requests
will fail and will not wait sleeping.

rafaeljusto commented on db/db.go:
```diff
@@ -127,6 +132,54 @@ func NewTx(db *sql.DB, timeout time.Duration) (*sql.Tx, error) {
+	}
+
+	d.Start()
+	for range time.Tick(2 * time.Second) {
```
The `2 * time.Second` should be replaced by the timeout used in `NewTx` function.

rafaeljusto commented on db/db.go:
```diff
@@ -127,6 +132,54 @@ func NewTx(db *sql.DB, timeout time.Duration) (*sql.Tx, error) {
+	checking bool
+}
+
+func (d *dbChecker) Checking() bool {
```
All methods from this dbChecker type could also be private (lowercase).

rafaeljusto commented on db/db.go:
```diff
@@ -127,6 +132,54 @@ func NewTx(db *sql.DB, timeout time.Duration) (*sql.Tx, error) {
+
+	d.Start()
+	for range time.Tick(2 * time.Second) {
+		tx, err := db.Begin()
```
How long are we going to wait for the `Begin` to return something?

rafaeljusto commented on db/db.go:
```diff
@@ -127,6 +132,54 @@ func NewTx(db *sql.DB, timeout time.Duration) (*sql.Tx, error) {
+
+		// db is back!
+		d.Stop()
+		tx.Commit()
```
Why do we commit here? Isn't this going to cause an error of using a transaction that was already committed?

rafaeljusto commented:
Unit tests are failing.

```
--- FAIL: ExampleNewTx (0.00s)
got:
new transaction timed out
want:
true
FAIL
```

Lets try to keep the test coverage in 100%. :smile:

rafaeljusto commented on db/db.go:
```diff
@@ -100,6 +101,10 @@ func ConnectPostgres(d ConnParams) (db *sql.DB, err error) {
 // NewTx starts a new database transaction with a timeout. Starting a new transaction is not
 // supposed to last too long, so a timeout of more than 3 seconds it's usually not necessary.
 func NewTx(db *sql.DB, timeout time.Duration) (*sql.Tx, error) {
```
We should improve the comments here explaining the fail-fast approach.

joaopetreli replied:
I think that would be good if the package documentation describes the fail-fast, and here there would be just a note or a short description.

joaopetreli replied:
Also the package documentation should mention that this fail-fast feature requires no caller action, it is done even though the caller do not need it.

Commit:
This implements some fixes pointed by @rafaeljusto

joaopetreli commented on db/db.go:
```diff
@@ -144,39 +152,39 @@ type dbChecker struct {
 }
 
-func (d *dbChecker) Checking() bool {
+func (d *dbChecker) checking() bool {
```
Should this method be named as Unreachable? It's seems that checking name does not pass the right idea, the idea that the database is unreachable.

And for flexibility purpose this method may be public, so clients of this library can call it where they think it's better.

rafaeljusto commented on db/db.go:
```diff
@@ -100,6 +101,21 @@ func ConnectPostgres(d ConnParams) (db *sql.DB, err error) {
 // NewTx starts a new database transaction with a timeout. Starting a new transaction is not
 // supposed to last too long, so a timeout of more than 3 seconds it's usually not necessary.
 func NewTx(db *sql.DB, timeout time.Duration) (*sql.Tx, error) {
+	if checker.checking() {
```
This approach assumes that the API user uses only one `sql.DB`, because it controls the unreachable database case with a global controller. Maybe a better idea is to make the global variable `checker` a map of `sql.DB` pointer to `dbChecker`.

rafaeljusto commented:
Patch proposal for this PR:

``` diff
diff --git a/db/db.go b/db/db.go
index c71586b..71c6480 100644
--- a/db/db.go
+++ b/db/db.go
@@ -18,8 +18,14 @@ import (
        "time"
 )

-// ErrNewTxTimedOut returned when takes too long to retrieve a transaction from the database server.
-var ErrNewTxTimedOut = errors.New("new transaction timed out")
+var (
+       // ErrNewTxTimedOut returned when takes too long to retrieve a transaction
+       // from the database server.
+       ErrNewTxTimedOut = errors.New("new transaction timed out")
+
+       // ErrUnreachable returned when the database is current unreachable.
+       ErrUnreachable = errors.New("unreachable database")
+)

 // PostgresDriver holds the postgres driver name. You should import the driver somewhere in your
 // code. The default value for PostgresDriver was get from https://github.com/lib/pq. This driver
@@ -101,14 +107,22 @@ func ConnectPostgres(d ConnParams) (db *sql.DB, err error) {
 // NewTx starts a new database transaction with a timeout. Starting a new transaction is not
 // supposed to last too long, so a timeout of more than 3 seconds it's usually not necessary.
 func NewTx(db *sql.DB, timeout time.Duration) (*sql.Tx, error) {
-       if checker.checking() {
-               return nil, ErrNewTxTimedOut // TODO change error type (?)
+       checkersLock.Lock()
+       checker, ok := checkers[db]
+       if !ok {
+               checker = new(dbChecker)
+               checkers[db] = checker
+       }
+       checkersLock.Unlock()
+
+       if checker.unreachable() {
+               return nil, ErrUnreachable
        }

        tx, err := newTx(db, timeout)
        if err == ErrNewTxTimedOut {
                go func() {
-                       checker.Check(db, timeout)
+                       checker.check(db, timeout)
                }()
        }

@@ -147,12 +161,26 @@ func newTx(db *sql.DB, timeout time.Duration) (*sql.Tx, error) {
        }
 }

+// Unreachable fast way to check if a database connection is unreachable without
+// creating a new transaction.
+func Unreachable(db *sql.DB) bool {
+       checkersLock.Lock()
+       defer checkersLock.Unlock()
+
+       if checker, ok := checkers[db]; ok {
+               return checker.unreachable()
+       }
+
+       // non connection created yet, so we assume that the database is reachable.
+       return false
+}
+
 type dbChecker struct {
        sync.RWMutex
        checking bool
 }

-func (d *dbChecker) checking() bool {
+func (d *dbChecker) unreachable() bool {
        d.RLock()
        defer d.RUnlock()
        return d.checking
@@ -170,8 +198,8 @@ func (d *dbChecker) stop() {
        d.checking = false
 }

-func (d *dbChecker) Check(db *sql.DB, duration time.Duration) {
-       if d.checking() {
+func (d *dbChecker) check(db *sql.DB, duration time.Duration) {
+       if d.unreachable() {
                // already checking
                return
        }
@@ -190,4 +218,7 @@ func (d *dbChecker) Check(db *sql.DB, duration time.Duration) {
        }
 }

-var checker = new(dbChecker)
+var (
+       checkers     = make(map[*sql.DB]*dbChecker)
+       checkersLock sync.Mutex
+)
```

rafaeljusto commented:
One problem raised by @trajber about my patch is that having a global locker would make it unscalable. We could improve it using a read-write lock.

A new idea was born from that: we could create a new layer over the sql.DB and add this fail-fast feature in it. The downside is that the API would start working with this specific type instead of the standard sql.DB.

rafaeljusto commented:
New patch after testing @trajber idea:

``` diff
diff --git a/db/db.go b/db/db.go
index c71586b..be88b30 100644
--- a/db/db.go
+++ b/db/db.go
@@ -18,8 +18,14 @@ import (
        "time"
 )

-// ErrNewTxTimedOut returned when takes too long to retrieve a transaction from the database server.
-var ErrNewTxTimedOut = errors.New("new transaction timed out")
+var (
+       // ErrNewTxTimedOut returned when takes too long to retrieve a transaction
+       // from the database server.
+       ErrNewTxTimedOut = errors.New("new transaction timed out")
+
+       // ErrUnreachable returned when the database is current unreachable.
+       ErrUnreachable = errors.New("unreachable database")
+)

 // PostgresDriver holds the postgres driver name. You should import the driver somewhere in your
 // code. The default value for PostgresDriver was get from https://github.com/lib/pq. This driver
@@ -46,6 +52,8 @@ type ConnParams struct {
        // StatementTimeout is the timeout utilized to invalidate queries that last more than the
        // configured timeout.
        StatementTimeout time.Duration
+       // TxTimeout is the timeout for creating a new database transaction
+       TxTimeout time.Duration

        MaxIdleConnections int
        MaxOpenConnections int
@@ -58,15 +66,24 @@ func NewConnParams() ConnParams {
                Host:               "127.0.0.1",
                ConnectTimeout:     3 * time.Second,
                StatementTimeout:   10 * time.Second,
+               TxTimeout:          3 * time.Second,
                MaxIdleConnections: 16,
                MaxOpenConnections: 32,
        }
 }

+// DB is a layer over the sql.DB to allow fail-fast approach when a database is
+// unreachable and to allow carrying the transaction timeout.
+type DB struct {
+       *sql.DB
+       checker   dbChecker
+       txTimeout time.Duration
+}
+
 // ConnectPostgres connects to a postgres database using the values from d. In case of a successfully
 // connection it returns a sql.DB and a nil error. In case of problem it returns a nil sql.DB and an
 // error from sql.Open (standard library, see https://golang.org/pkg/database/sql/#Open)
-func ConnectPostgres(d ConnParams) (db *sql.DB, err error) {
+func ConnectPostgres(d ConnParams) (*DB, error) {
        // connect_timeout
        //
        // https://www.postgresql.org/docs/9.6/static/libpq-connect.html#LIBPQ-CONNECT-CONNECT-TIMEOUT
@@ -89,26 +106,31 @@ func ConnectPostgres(d ConnParams) (db *sql.DB, err error) {
                int(d.StatementTimeout.Seconds()*1000),
        )

-       if db, err = sql.Open(PostgresDriver, connParams); err != nil {
+       db, err := sql.Open(PostgresDriver, connParams)
+       if err != nil {
                return nil, err
        }

        db.SetMaxIdleConns(d.MaxIdleConnections)
        db.SetMaxOpenConns(d.MaxOpenConnections)
-       return
+
+       return &DB{
+               DB:        db,
+               txTimeout: d.TxTimeout,
+       }, nil
 }

 // NewTx starts a new database transaction with a timeout. Starting a new transaction is not
 // supposed to last too long, so a timeout of more than 3 seconds it's usually not necessary.
-func NewTx(db *sql.DB, timeout time.Duration) (*sql.Tx, error) {
-       if checker.checking() {
-               return nil, ErrNewTxTimedOut // TODO change error type (?)
+func NewTx(db *DB) (*sql.Tx, error) {
+       if db.checker.unreachable() {
+               return nil, ErrUnreachable
        }

-       tx, err := newTx(db, timeout)
+       tx, err := newTx(db.DB, db.txTimeout)
        if err == ErrNewTxTimedOut {
                go func() {
-                       checker.Check(db, timeout)
+                       db.checker.check(db.DB, db.txTimeout)
                }()
        }

@@ -147,12 +169,18 @@ func newTx(db *sql.DB, timeout time.Duration) (*sql.Tx, error) {
        }
 }

+// Unreachable fast way to check if a database connection is unreachable without
+// creating a new transaction.
+func Unreachable(db *DB) bool {
+       return db.checker.unreachable()
+}
+
 type dbChecker struct {
        sync.RWMutex
        checking bool
 }

-func (d *dbChecker) checking() bool {
+func (d *dbChecker) unreachable() bool {
        d.RLock()
        defer d.RUnlock()
        return d.checking
@@ -170,8 +198,8 @@ func (d *dbChecker) stop() {
        d.checking = false
 }

-func (d *dbChecker) Check(db *sql.DB, duration time.Duration) {
-       if d.checking() {
+func (d *dbChecker) check(db *sql.DB, duration time.Duration) {
+       if d.unreachable() {
                // already checking
                return
        }
@@ -189,5 +217,3 @@ func (d *dbChecker) Check(db *sql.DB, duration time.Duration) {
                return
        }
 }
-
-var checker = new(dbChecker)
diff --git a/db/db_test.go b/db/db_test.go
index 0118984..9963aca 100644
--- a/db/db_test.go
+++ b/db/db_test.go
@@ -1,7 +1,6 @@
 package db_test

 import (
-       "database/sql"
        "database/sql/driver"
        "fmt"
        "reflect"
@@ -58,6 +57,10 @@ func TestConnectPostgres(t *testing.T) {
                db.PostgresDriver = originalPostgresDriver
        }()

+       defer func() {
+               testdb.SetOpenFunc(nil)
+       }()
+
        for i, scenario := range scenarios {
                testdb.SetOpenFunc(scenario.openFunc)
                db.PostgresDriver = scenario.postgresDriver
@@ -77,62 +80,110 @@ func TestConnectPostgres(t *testing.T) {

 func TestNewTx(t *testing.T) {
        scenarios := []struct {
-               description   string
-               db            *sql.DB
-               beginFunc     func() (driver.Tx, error)
-               timeout       time.Duration
-               expectedError error
+               description    string
+               connParams     db.ConnParams
+               postgresDriver string
+               openFunc       func(dsn string) (driver.Conn, error)
+               beginFunc      func() (driver.Tx, error)
+               expectedError  error
        }{
                {
                        description: "it should initialize a transaction correctly",
-                       db: func() *sql.DB {
-                               db, err := sql.Open("testdb", "")
-                               if err != nil {
-                                       t.Fatal(err)
-                               }
-                               return db
+                       connParams: func() db.ConnParams {
+                               d := db.NewConnParams()
+                               d.DatabaseName = "test"
+                               d.Username = "user"
+                               d.Password = "passwd"
+                               d.TxTimeout = 1 * time.Second
+                               return d
                        }(),
+                       postgresDriver: "testdb",
+                       openFunc: func(dsn string) (driver.Conn, error) {
+                               connParams := "user=user password=passwd dbname=test sslmode=disable host=127.0.0.1 connect_timeout=3 statement_timeout=10000"
+                               if dsn != connParams {
+                                       return nil, fmt.Errorf("invalid connection string")
+                               }
+
+                               return testdb.Conn(), nil
+                       },
                        beginFunc: func() (driver.Tx, error) {
                                return &testdb.Tx{}, nil
                        },
-                       timeout: 1 * time.Second,
                },
                {
                        description: "it should detect an error while initializing a transaction",
-                       db: func() *sql.DB {
-                               db, err := sql.Open("testdb", "")
-                               if err != nil {
-                                       t.Fatal(err)
-                               }
-                               return db
+                       connParams: func() db.ConnParams {
+                               d := db.NewConnParams()
+                               d.DatabaseName = "test"
+                               d.Username = "user"
+                               d.Password = "passwd"
+                               d.TxTimeout = 1 * time.Second
+                               return d
                        }(),
+                       postgresDriver: "testdb",
+                       openFunc: func(dsn string) (driver.Conn, error) {
+                               connParams := "user=user password=passwd dbname=test sslmode=disable host=127.0.0.1 connect_timeout=3 statement_timeout=10000"
+                               if dsn != connParams {
+                                       return nil, fmt.Errorf("invalid connection string")
+                               }
+
+                               return testdb.Conn(), nil
+                       },
                        beginFunc: func() (driver.Tx, error) {
                                return nil, fmt.Errorf("i'm a crazy error")
                        },
-                       timeout:       1 * time.Second,
                        expectedError: fmt.Errorf("i'm a crazy error"),
                },
                {
                        description: "it should timeout when transaction takes too long to start",
-                       db: func() *sql.DB {
-                               db, err := sql.Open("testdb", "")
-                               if err != nil {
-                                       t.Fatal(err)
-                               }
-                               return db
+                       connParams: func() db.ConnParams {
+                               d := db.NewConnParams()
+                               d.DatabaseName = "test"
+                               d.Username = "user"
+                               d.Password = "passwd"
+                               d.TxTimeout = 10 * time.Millisecond
+                               return d
                        }(),
+                       postgresDriver: "testdb",
+                       openFunc: func(dsn string) (driver.Conn, error) {
+                               connParams := "user=user password=passwd dbname=test sslmode=disable host=127.0.0.1 connect_timeout=3 statement_timeout=10000"
+                               if dsn != connParams {
+                                       return nil, fmt.Errorf("invalid connection string")
+                               }
+
+                               return testdb.Conn(), nil
+                       },
                        beginFunc: func() (driver.Tx, error) {
                                time.Sleep(1 * time.Second)
                                return &testdb.Tx{}, nil
                        },
-                       timeout:       10 * time.Millisecond,
                        expectedError: db.ErrNewTxTimedOut,
                },
        }

+       originalPostgresDriver := db.PostgresDriver
+       defer func() {
+               db.PostgresDriver = originalPostgresDriver
+       }()
+
+       defer func() {
+               testdb.SetOpenFunc(nil)
+               testdb.SetBeginFunc(nil)
+       }()
+
        for i, scenario := range scenarios {
+               testdb.SetOpenFunc(scenario.openFunc)
                testdb.SetBeginFunc(scenario.beginFunc)
-               tx, err := db.NewTx(scenario.db, scenario.timeout)
+               db.PostgresDriver = scenario.postgresDriver
+
+               d, err := db.ConnectPostgres(scenario.connParams)
+               if err != nil {
+                       t.Errorf("scenario %d, “%s”: error connecting database. Details: %s",
+                               i, scenario.description, err)
+                       continue
+               }
+
+               tx, err := db.NewTx(d)

                if scenario.expectedError == nil && tx == nil {
                        t.Errorf("scenario %d, “%s”: tx not initialized",
@@ -156,6 +207,7 @@ func ExampleConnectPostgres() {
                Host:               "localhost:5432",
                ConnectTimeout:     3 * time.Second,
                StatementTimeout:   10 * time.Second,
+               TxTimeout:          3 * time.Second,
                MaxIdleConnections: 16,
                MaxOpenConnections: 32,
        }
@@ -182,6 +234,7 @@ func ExampleNewTx() {
                Host:               "localhost:5432",
                ConnectTimeout:     3 * time.Second,
                StatementTimeout:   10 * time.Second,
+               TxTimeout:          3 * time.Second,
                MaxIdleConnections: 16,
                MaxOpenConnections: 32,
        }
@@ -193,7 +246,7 @@ func ExampleNewTx() {
                return
        }

-       tx, err := db.NewTx(dbConn, 3*time.Second)
+       tx, err := db.NewTx(dbConn)
        if err != nil {
                fmt.Println(err)
                return
```

rafaeljusto committed:
Allow fail-fast for multiple database connections

Refactory to split the connection initialization from the transaction timeout
with fail-fast feature. Now the API will work with its own DB type to allow a
lock free fail-fast connection.

rafaeljusto commented:
We still need to improve the tests to cover the fail-fast cases before merging this PR.

rafaeljusto committed:
Avoid more than 1 checker running at the same time

rafaeljusto committed:
Improve test coverage
